runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java

```diff
@@ -275,13 +277,53 @@ private boolean matches(MetricsFilter filter, MetricKey key) {
         && matchesScope(key.stepName(), filter.steps());
   }
 
-  private boolean matchesScope(String actualScope, Set<String> scopes) {
+  /** Convert a list of Strings into a Map of String to its index in the list. */
+  private Map<String, List<Integer>> makeScopeIndex(String[] scopeSplit) {
+    Map<String, List<Integer>> result = new HashMap<String, List<Integer>>();
+    for (int i = 0; i < scopeSplit.length; i++) {
+      String stage = scopeSplit[i];
+      if (!result.containsKey(stage)) {
+        result.put(stage, new ArrayList<Integer>());
+      }
+      result.get(stage).add(i);
+    }
+    return result;
+  }
+
+  /** Realize a substring search between the actual scope and the filter scope. */
+  private boolean matchesDeepScope(String[] actualScopeSplit,
+      Map<String, List<Integer>> actualScopeIndex, String filterScope) {
+    String[] splitScope = filterScope.split("/");
+    if (!actualScopeIndex.containsKey(splitScope[0])) {
+      return false;
+    }
+    for (int startingIndex : actualScopeIndex.get(splitScope[0])) {
+      int i = 0;
+      while (i < splitScope.length && startingIndex + i < actualScopeSplit.length) {
+        if (!splitScope[i].equals(actualScopeSplit[startingIndex + i])) {
+          break;
+        }
+        i++;
+      }
+      if (i == splitScope.length) {
+        return true;
+      }
+    }
+    return false;
+  }
+  /** Check if the scope of a metric is matched by any of the filters. */
```

Comment:
This should indicate what constitutes "a match", especially on scopes. Also, newline between previous method and comments for this method.